garage/misc/git_handler.py

```diff
@@ -0,0 +1,382 @@
+                colored("The reference %s does not exist", "red") % reference)
+            raise
+        return sha
+
+    def create_stash(self):
+        """Stashes the local changes that have not been staged.
+
+        The stash is not stored anywhere in the ref namespace so the user does
+        not have remove it manually later.
+
+        Returns
+        -------
+        The 40 character SHA if local changes were found, or an empty string
+        otherwise.
+
+        """
+        stash_sha = ""
+        try:
+            diff_cmd = self.git.diff("--exit-code")
+        except sh.ErrorReturnCode_1:
+            # Exit code 1 means there are differences
+            create_stash_cmd = self.git.stash.create()
+            stash_sha = cmd_to_string(create_stash_cmd)
+            pass
+        return stash_sha
+
+    def get_current_branch_name(self):
+        """Return the name of the checked out branch.
+
+        If the HEAD is in detached state, an empty string is returned.
+
+        Precondition
+        ------------
+        - No branch (local or remote) in the repository is named "HEAD".
+
+        """
+        rev_cmd = self.git("rev-parse", "--abbrev-ref", "HEAD")
+        curr_branch_name = cmd_to_string(rev_cmd)
+        if curr_branch_name == "HEAD":
+            curr_branch_name = ""
+        return curr_branch_name
+
+    def run_experiment_on_target(self, target, exp_args):
+        """Run a experiment on the indicated commit defined by target.
+
+        After the target is verified, if there are any non-staged changes in
+        the working directory, they're stashed, the target commit is checked
+        out, the experiment is executed and the working directory is restored
+        at the end, including the stashed changes.
+
+        Parameters
+        ----------
+        - target: a key-value string, where the key indicates the type of
+            target (tag, branch, sha), and the value the reference for the
+            corresponding type. For example:
+                sha: 8d54s23a
+                branch: my_branch
+                branch: remote/master
+                tag: my_tag
+        - exp_args: it's a dictionary of arguments obtained from the function
+            run_experiment at garage.misc.instrument, that will be passed to
+            the experiment once the target is checked out.
+
+        Precondition
+        ------------
+        - The working directory is not in the middle of a conflict resolution.
+
+        """
+        assert isinstance(target, str), ("The git target has to be a string")
+        self.target = target
+        self.target_sha = self._verify_target(self.target)
+        self.restore_sha = self.get_sha()
+        self.stash_sha = self.create_stash()
+        self.branch_name = self.get_current_branch_name()
+        self.target_branch_name = ("run_exp_on_" + str(int(time.time())))
+        try:
+            self._switch_to_target(self.target_branch_name, self.target_sha)
```

Comment:
It's true that we can use the same local branch or detached HEAD to switch between commits, but in the case of local branches, if an unexpected error happens and not even this [code](https://github.com/rlworkgroup/garage/blob/1e6d931f4fc163040b859450db769f7049365527/garage/misc/git_handler.py#L285) restores the branch correctly, I think it's less risky to not lose the reference of local commits under the branch by creating a new one just for the test.
It's also true that developers can use `git reflog` to look for these lost references, but maybe for someone not so experienced in git it could be easier just to have the reference of the original branch and delete the branch used for testing.

Reply:
Sounds good!